### PR TITLE
fix(SdkList):  fix list delete values panic

### LINKED ISSUE #3614: [Lists]: list(empty) delete values will  be panic
### SDK(sidecar)Panic 💣  

**How to reproduce it (as minimally and precisely as possible)**:
 * [FeatureFlag:CountsAndLists]
    remove `Lists` values when  `Lists` empty, SDK(sidecar) will panic   

**Environment**:
- Agones version: 1.37.0
- Kubernetes version (use `kubectl version`): 1.29
- Install method (yaml/helm):  helm
- Troubleshooting guide log(s):      
 ```  text
E0126 12:23:35.015025       1 runtime.go:79] Observed a panic: "makeslice: cap out of range" (runtime error: makeslice: cap out of range)
goroutine 61 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x1bcfb80?, 0x22d81d0})
        /go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:75 +0x99
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc000061d30?})
        /go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:49 +0x75
panic({0x1bcfb80, 0x22d81d0})
        /usr/local/go/src/runtime/panic.go:884 +0x213
agones.dev/agones/pkg/sdkserver.deleteValues(...)
        /go/src/agones.dev/agones/pkg/sdkserver/sdkserver.go:1241
agones.dev/agones/pkg/sdkserver.(*SDKServer).updateList(0xc00077a1e0, {0x22f8000, 0xc0000ac910})
        /go/src/agones.dev/agones/pkg/sdkserver/sdkserver.go:1205 +0x40f
agones.dev/agones/pkg/sdkserver.(*SDKServer).syncGameServer(0x0?, {0x22f8000?, 0xc0000ac910?}, {0x1e90d8c?, 0xffffffffffffffff?})
        /go/src/agones.dev/agones/pkg/sdkserver/sdkserver.go:353 +0x89
agones.dev/agones/pkg/util/workerqueue.(*WorkerQueue).processNextWorkItem(0xc0000ac730, {0x22f8000, 0xc0000ac910})
        /go/src/agones.dev/agones/pkg/util/workerqueue/workerqueue.go:182 +0x274
agones.dev/agones/pkg/util/workerqueue.(*WorkerQueue).runWorker(...)
        /go/src/agones.dev/agones/pkg/util/workerqueue/workerqueue.go:158
agones.dev/agones/pkg/util/workerqueue.(*WorkerQueue).run.func1()
        /go/src/agones.dev/agones/pkg/util/workerqueue/workerqueue.go:217 +0x45
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x30?)
        /go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:226 +0x3e
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x17aeb16?, {0x22dfb80, 0xc000894510}, 0x1, 0xc000542ea0)
        /go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:227 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x0?, 0x3b9aca00, 0x0, 0x0?, 0x0?)
        /go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:204 +0x89
k8s.io/apimachinery/pkg/util/wait.Until(...)
        /go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:161
agones.dev/agones/pkg/util/workerqueue.(*WorkerQueue).run(0xc0000ac730?, {0x22f8000?, 0xc0000ac910?})
        /go/src/agones.dev/agones/pkg/util/workerqueue/workerqueue.go:217 +0xdc
created by agones.dev/agones/pkg/util/workerqueue.(*WorkerQueue).Run
        /go/src/agones.dev/agones/pkg/util/workerqueue/workerqueue.go:206 +0x165
panic: runtime error: makeslice: cap out of range [recovered]
        panic: runtime error: makeslice: cap out of range
```

### PR DESCRIPTION
**Pull request by GStones**
**What type of PR is this?**
/kind bug

**What this PR does / Why we need it**:
 It will make Sdk sidecar panic
**Which issue(s) this PR fixes**:

Closes #3614

### CODE REVIEW

**Comment by jlory** [comment]
I saw that issue as well, as you said it's fixed by the other mr, I would say that it's still dangerous to have possible negative capacity so it should be fixed however I'm not sure there is a real performance improvement  by trying to prealocate the map.

**Comment by GStones** [comment]
> I saw that issue as well, as you said it's fixed by the other mr, I would say that it's still dangerous to have possible negative capacity so it should be fixed however I'm not sure there is a real performance improvement  by trying to prealocate the map.

Here is preallocate a slice, I think it can be removed,because the list max available is 1000, slice grow has limit(1024).

File: pkg/sdkserver/sdkserver.go [hunk 1]
```diff
@@ -1238,13 +1239,14 @@ func (s *SDKServer) updateList(ctx context.Context) error {
 
 // Returns a new string list with the string keys in toDeleteValues removed from valuesList.
 func deleteValues(valuesList []string, toDeleteValues map[string]bool) []string {
-	newList := make([]string, 0, len(valuesList)-len(toDeleteValues))
-	for i, val := range valuesList {
-		if _, ok := toDeleteValues[val]; !ok {
-			newList = append(newList, valuesList[i])
+	newValuesList := make([]string, 0, len(valuesList))
```
Comment (markmandel) [inline pkg/sdkserver/sdkserver.go:1242]: Possibly easier suggestion for defensive programming.

```suggestion
  newLen = len(valuesList)-len(toDeleteValues)
  if newLen < 0 {
    newLen = 0
  }

	newList := make([]string, 0, newLen)
```

And then we continue on as before.

Or I think as you said, make `newValueList := []string{}` and let it expand naturally - which is a bit more readable IMHO, but I don't have strong opinions about either approach.

Comment (GStones) [reply pkg/sdkserver/sdkserver.go]: Thanks for your suggestions!  Already fixed, please review

**Comment by markmandel** [comment]
I see this failure in the unit tests:

```
--- FAIL: TestSDKServerRemoveListValue (28.53s)
    --- FAIL: TestSDKServerRemoveListValue/Remove_all_values (1.11s)
        sdkserver_test.go:1618: 
            	Error Trace:	/go/src/agones.dev/agones/pkg/sdkserver/sdkserver_test.go:1618
            	Error:      	Not equal: 
            	            	expected: []string{}
            	            	actual  : []string(nil)
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,3 +1,2 @@
            	            	-([]string) {
            	            	-}
            	            	+([]string) <nil>
            	            	 
            	Test:       	TestSDKServerRemoveListValue/Remove_all_values
        sdkserver_test.go:1626: 
            	Error Trace:	/go/src/agones.dev/agones/pkg/sdkserver/sdkserver_test.go:1626
            	Error:      	Not equal: 
            	            	expected: v1.ListStatus{Capacity:100, Values:[]string{}}
            	            	actual  : v1.ListStatus{Capacity:100, Values:[]string(nil)}
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -2,4 +2,3 @@
            	            	  Capacity: (int64) 100,
            	            	- Values: ([]string) {
            	            	- }
            	            	+ Values: ([]string) <nil>
            	            	 }
            	Test:       	TestSDKServerRemoveListValue/Remove_all_values
```

Do you see the same thing locally, or is this a flake?

**Comment by markmandel** [comment]
```
--- FAIL: TestSDKServerAddListValue (31.96s)
    --- FAIL: TestSDKServerAddListValue/Add_multiple_values_past_capacity (1.12s)
        sdkserver_test.go:1477: 
            	Error Trace:	/go/src/agones.dev/agones/pkg/sdkserver/sdkserver_test.go:1477
            	Error:      	Not equal: 
            	            	expected: v1.ListStatus{Capacity:10, Values:[]string{"one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "ten"}}
            	            	actual  : v1.ListStatus{Capacity:10, Values:[]string{"one", "two", "three", "four", "five", "six", "seven", "eight", "nine"}}
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -2,3 +2,3 @@
            	            	  Capacity: (int64) 10,
            	            	- Values: ([]string) (len=10) {
            	            	+ Values: ([]string) (len=9) {
            	            	   (string) (len=3) "one",
            	            	@@ -11,4 +11,3 @@
            	            	   (string) (len=5) "eight",
            	            	-  (string) (len=4) "nine",
            	            	-  (string) (len=3) "ten"
            	            	+  (string) (len=4) "nine"
            	            	  }
            	Test:       	TestSDKServerAddListValue/Add_multiple_values_past_capacity
        sdkserver_test.go:1487: 
            	Error Trace:	/go/src/agones.dev/agones/pkg/sdkserver/sdkserver_test.go:1487
            	Error:      	Not equal: 
            	            	expected: 0
            	            	actual  : 1
            	Test:       	TestSDKServerAddListValue/Add_multiple_values_past_capacity
```

Looks like this flake is hitting us hard. We should dig into this one @igooch

**Comment by markmandel** [comment]
Okay I can at least replicate this! It's going to only happen under load.

So to replicate, we drop in a `time.Sleep(time.Second)` to replicate load in the for loop.

```
			for i, req := range testCase.requests {
				_, err = sc.AddListValue(context.Background(), req)
				if testCase.updateErrs[i] {
					assert.Error(t, err)
				} else {
					assert.NoError(t, err)
				}
				time.Sleep(time.Second)
			}
			
```

/cc @igooch in case you see something I'm missing.

**Comment by markmandel** [comment]
Got it! Fixed. PR incoming shortly.

**Comment by markmandel** [comment]
Whoops, was writing comments on the wrong issue! 🤦🏻 

Either way #3627 will fix the flake.

**Review by markmandel — APPROVED** [review]

**Pull request merged**